explainer.md

```diff
@@ -0,0 +1,208 @@
+# WebXR Device API - Hand Input
+
+This document describes a design giving developers access to hand-tracking XR systems, building on top of the [WebXR device API](https://immersive-web.github.io/webxr/)
+
+## Use cases and scope
+
+This API primarily exposes the poses of hand skeleton joints. It can be used to render a hand model in VR scenarios, as well perform gesture detection with the hands. It does not provide access to a full hand mesh.
+
+## Accessing this API
+
+This API will only be accessible if a `"hand-tracking"` [XR feature](https://immersive-web.github.io/webxr/#feature-dependencies) is requested.
+
+This API presents itself as an additional field on `XRInputSource`, `hand`. The `hand` attribute will be non-null if the input source supports hand tracking and the feature has been requested.
+
+```js
+navigator.xr.requestSession({optionalFeatures: ["hand-tracking"]}).then(...);
+
+function renderFrame(session, frame) {
+   // ...
+
+   for (inputSource of session.inputSources) {
+      if (inputSource.hand) {
+         // render a hand model
+         // perform gesture detection
+      }
+   }
+}
+
+
+```
+
+## Hands and joints
+
+Each hand is made up many bones, connected by _joints_. We name them with their connected bone, for example `INDEX_PHALANGE_DISTAL` is the joint closer to the wrist connected to the distal phalange bone of the index finger. The `*_PHALANGE_TIP` "joints" locate the tips of the fingers. The `WRIST` joint is located at the composite joint between the wrist and forearm.
+
+The joints can be accessed via indexing, for example to access the middle knuckle joint one would use:
```

Comment:
I'm curious: What's the benefit to using indexing in this case vs simply having list of named attributes? ie:

```webidl
partial interface XRHand {
  XRJoint? wrist;
  XRJoint? thumbTrapezium;
  XRJoint? thumbPhalangeProximal ;
 // etc...
}
```

Reply:
Actually, as I finish typing that I can think of a couple, but I'm curious to hear the authors intent. :)

Reply:
one of the examples below demonstrates it, by using indexing we can do things like "iterate over all joints on the index finger, in order". Now, we could also add first class APIs for that (`getIndexJoints()`, etc), however in that case you would still want to know which joint you got, so you _still_ need those constants (or an enum) for telling them apart. I do not have a _strong_ preference for this model, but I find it to be overall cleaner. I think it would potentially be nice to _also_ have higher level APIs on top of this.

Reply:
I can see both sides of this argument.  As it stands, the OpenXR hand proposal is evolving a bit to allow easier enumeration over the agreed-on common set of joints, and so I think I weakly prefer the proposal as is, even though it produces a bit uglier code than the alternative.

I was going to ask if there was much web precedent for an API that defines a virtual enum of manually-numbered constants like this, but then I reminded myself of WebGL :smile:  However, there I presume the intention was compatibility with OpenGL ES API shape - is TAG on board with this API approach?

Reply:
manually numbered constants exist all over the place, [there are two sets on Node](https://dom.spec.whatwg.org/#node), and some on [`CSSRule`](https://drafts.csswg.org/cssom/#cssrule), [WebSocket](`https://html.spec.whatwg.org/multipage/web-sockets.html#websocket`), [`DOMException`](https://heycam.github.io/webidl/#idl-DOMException), [`XMLHttpRequest`](https://xhr.spec.whatwg.org/#xmlhttprequest), and those are just the ones I remember.

I don't mind making this an enum instead, I just kinda liked the constants+indexing model and I've seen lots of specs use constants like this before.

Reply:
The current approach works for me - just wanted to be sure it was "webby" and that we wouldn't be asked to change it later!

Reply:
I think we're far too early in the process to need to worry about that :smile: , this is easy enough to change and I don't think we'll be adding much more to _this_ side of the API for a while anyway. We should definitely get the TAG to check it before embarking on any major addition to how we identify joints (e.g. if we add an API that separates out fingers, etc)